---
title: "Data preparation and processing"
output: html_notebook
---

## 0.
scRNA-seq data from the transcriptomes of 6 mouse spinal cords, sequenced in Blum 21. Download data from:
路 Alpha:            http://spinalcordatlas.org/assets/downloads/alphamns.h5ad
路 Gamma:            http://spinalcordatlas.org/assets/downloads/gammamns.h5ad
路 Visceral:         http://spinalcordatlas.org/assets/downloads/visceralmns.h5ad
路 Whole dataset:    http://spinalcordatlas.org/assets/downloads/allexpsctl.h5ad

## 1. 
Import libraries, create directories for saving images

```{r}
library(Seurat)
library(zellkonverter)
library(ggplot2)
library(SummarizedExperiment)
library(SingleCellExperiment)
library(magrittr)
library(scPred)

rds.save.dir <- "/Users/xuanzhao/Documents/R/PredictResistance/"
graph.save.dir <- "/Users/xuanzhao/Documents/R/PredictResistance/graphs/"
```

## 2.
Load data (h5ad files) as objects of class SingleCellExperiment

```{r}
alpha.data <- readH5AD(file = "~/Documents/R/GSE161621_filtered_data/alphamns.h5ad")
gamma.data <- readH5AD(file = "~/Documents/R/GSE161621_filtered_data/gammamns.h5ad")
visceral.data <- readH5AD(file = "~/Documents/R/GSE161621_filtered_data/visceralmns.h5ad")
```

## 3.
Convert objects to class Seurat

```{r}
# alpha MNs
a.counts <- assay(alpha.data, "X")
a.libsizes <- colSums(a.counts)
a.size.factors <- a.libsizes/mean(a.libsizes)
logcounts(alpha.data) <- log2(t(t(a.counts)/a.size.factors) + 1)            # calculate + add log-transformed normalized expression values from the count matrix 
logcounts(alpha.data) = as.matrix(logcounts(alpha.data))                      # convert logcounts sparse matrix to matrix
alpha <- as.Seurat(alpha.data, counts = "X", data = "logcounts")  # convert to seurat object

# gamma MNs
g.counts <- assay(gamma.data, "X")
g.libsizes <- colSums(g.counts)
g.size.factors <- g.libsizes/mean(g.libsizes)
logcounts(gamma.data) <- log2(t(t(g.counts)/g.size.factors) + 1)
logcounts(gamma.data) = as.matrix(logcounts(gamma.data))
gamma <- as.Seurat(gamma.data, counts = "X", data = "logcounts")

# visceral MNs
v.counts <- assay(visceral.data, "X")
v.libsizes <- colSums(v.counts)
v.size.factors <- v.libsizes/mean(v.libsizes)
logcounts(visceral.data) <- log2(t(t(v.counts)/v.size.factors) + 1)
logcounts(visceral.data) = as.matrix(logcounts(visceral.data))
visceral <- as.Seurat(visceral.data, counts = "X", data = "logcounts")
```

## 4.
Split alpha MNs into fast-firing, fast-fatigue-resistant, and slow-firing (ff, fr, sf)

```{r}
DefaultAssay(alpha) <- 'RNA' # default assay set to RNA

Idents(object = alpha, WhichCells(object = alpha, expression = Prkcd>0 & Sv2a>0 & Chodl==0, slot = 'data')) <- "sf"

Idents(object = alpha, WhichCells(object = alpha, expression = Chodl>0 & Kcnq5>0 & Sv2a==0, slot = 'data')) <- "ff"

Idents(object = alpha, WhichCells(object = alpha, expression = Chodl>0 & Kcnq5==0 & Sv2a==0, slot = 'data')) <- "fr"

alpha <- subset(x = alpha, idents = "SingleCellExperiment", invert = TRUE) # removes ident SingleCellExperiment
```

## 5.
Split gamma MNs into gamma and beta by ident

```{r}
beta.subset <- subset(x = gamma, subset = skeletal_subtype == 'Beta(?) Motor Neurons')
Idents(object = gamma, WhichCells(object = gamma, cells = colnames(beta.subset))) <- 'beta'

gamma.subset <- subset(x = gamma, subset = skeletal_subtype == 'Gamma Motor Neurons')
Idents(object = gamma, WhichCells(object = gamma, cells = colnames(gamma.subset))) <- 'gamma'
```

## 6.
Combine datasets

```{r}
ag.combined <- merge(x = alpha, y = gamma)
all.genes <- rownames(ag.combined)

ag.combined <- ag.combined %>%
  NormalizeData() %>%
  FindVariableFeatures(selection.method = "vst", nfeatures = 2000) %>%
  ScaleData(features = all.genes) %>%
  RunPCA(features = VariableFeatures(object = ag.combined)) %>%
  RunUMAP(dims = 1:10) # dims = 1:30 instead?

# Error in irlba(A = t(x = object), nv = npcs, ...) : max(nu, nv) must be positive
# Try running the code into the console without magrittr %>% pipes
```

## 7.
Add metadata "phenotype," which stores each training cell's classification as "res" or "vuln"

```{r}
# pbmc$CellType <- Idents(pbmc)

# Create 'phenotype' dataframe (# of rows = # of rows in ag.combined)
phenotype <- data.frame(rownames(ag.combined@meta.data), row.names = rownames(ag.combined@meta.data))
colnames(phenotype) <- "p"

# Subsets of ff, fr, sf, b, and g data
ff.subset <- subset(ag.combined, idents = "ff")
fr.subset <- subset(ag.combined, idents = "fr")
sf.subset <- subset(ag.combined, idents = "sf")
# beta.subset
# gamma.subset

ff.cellNames <- rownames(ff.subset@meta.data)
fr.cellNames <- rownames(fr.subset@meta.data)
vuln.cellNames <- c(ff.cellNames, fr.cellNames)

sf.cellNames <- rownames(sf.subset@meta.data)
g.cellNames <- rownames(gamma.subset@meta.data)
res.cellNames <- c(sf.cellNames, g.cellNames)

b.cellNames <- rownames(beta.subset@meta.data)

# Iteratively changes dataframe
phenotype$p[phenotype$p %in% vuln.cellNames] <- "vuln"
phenotype$p[phenotype$p %in% res.cellNames] <-  "res"
phenotype$p[phenotype$p %in% b.cellNames] <-  "beta_predict"

# Adds dataframe into ag.combined@meta.data - Can now be accessed by ag.combined$phenotype
ag.combined <- AddMetaData(object = ag.combined, metadata = phenotype, col.name = "phenotype")
```

## 8.
Take out beta cells from training data

```{r}
ag.training <- subset(ag.combined, idents = 'beta', invert = TRUE)
```

## 9.
General housekeeping

```{r}
# Amount of data per group

ag.combined.table <- table(Idents(ag.combined))  # table with number of cells in each identity

# Visualizing Kaplan 14 genes and their expression levels in this dataset
# Helps determine whether they would be a good fit for ML feature space confinement

AverageExpression(ag.combined, features = c("Sema3e","Gad2","Kcnj5","Rsph9","Npr3","Vipr2","Ecrg4","Egln3","Mgam","Arhgap9","Dlk1","Mmp9","Il1f9","Trhr","Aldh1a2","Cited1","Hsd17b2","Nts"), slot = 'data')

# Likely gene candidates:
# Sema3e (opposite), Gad2, Kcnj5, Arghap9, Dlk1 (maybe opposite), 
```


